src/java/org/apache/nutch/scoring/webgraph/LinkRank.java

```diff
@@ -350,186 +367,221 @@ public void close() {
-      if (numOutlinks > 0) {
-        for (int i = 0; i < outlinks.size(); i++) {
-          LinkDatum outlink = outlinks.get(i);
-          String toUrl = outlink.getUrl();
-
-          outlink.setUrl(fromUrl);
-          outlink.setScore(outlinkScore);
-
-          // collect the inverted outlink
-          output.collect(new Text(toUrl), outlink);
-          LOG.debug(toUrl + ": inverting inlink from " + fromUrl
-              + " origscore: " + inlinkScore + " numOutlinks: " + numOutlinks
-              + " inlinkscore: " + outlinkScore);
-        }
+      public void cleanup(){
       }
-    }
 
-    public void close() {
+      public void reduce(Text key, Iterable<ObjectWritable> values,
+          Context context)
+          throws IOException, InterruptedException {
+
+        String fromUrl = key.toString();
+        List<LinkDatum> outlinks = new ArrayList<>();
+        Node node = null;
+
+        // aggregate outlinks, assign other values
+        for (ObjectWritable write : values) {
+          Object obj = write.get();
+          if (obj instanceof Node) {
+            node = (Node) obj;
+          } else if (obj instanceof LinkDatum) {
+            outlinks.add(WritableUtils.clone((LinkDatum) obj, conf));
+          }
+        }
+
+        // get the number of outlinks and the current inlink and outlink scores
+        // from the node of the url
+        int numOutlinks = node.getNumOutlinks();
+        float inlinkScore = node.getInlinkScore();
+        float outlinkScore = node.getOutlinkScore();
+        LOG.debug(fromUrl + ": num outlinks " + numOutlinks);
+
+        // can't invert if no outlinks
+        if (numOutlinks > 0) {
+          for (int i = 0; i < outlinks.size(); i++) {
+            LinkDatum outlink = outlinks.get(i);
+            String toUrl = outlink.getUrl();
+
+            outlink.setUrl(fromUrl);
+            outlink.setScore(outlinkScore);
+
+            // collect the inverted outlink
+            context.write(new Text(toUrl), outlink);
+            LOG.debug(toUrl + ": inverting inlink from " + fromUrl
+                + " origscore: " + inlinkScore + " numOutlinks: " + numOutlinks
+                + " inlinkscore: " + outlinkScore);
+          }
+        }
+      }
     }
   }
 
   /**
    * Runs a single link analysis iteration.
    */
-  private static class Analyzer implements
-      Mapper<Text, Writable, Text, ObjectWritable>,
-      Reducer<Text, ObjectWritable, Text, Node> {
+  private static class Analyzer {
 
-    private JobConf conf;
-    private float dampingFactor = 0.85f;
-    private float rankOne = 0.0f;
-    private int itNum = 0;
-    private boolean limitPages = true;
-    private boolean limitDomains = true;
+    private static Configuration conf;
+    private static float dampingFactor = 0.85f;
+    private static float rankOne = 0.0f;
+    private static int itNum = 0;
+    private static boolean limitPages = true;
+    private static boolean limitDomains = true;
```

Comment:
See comments in Generator etc. regarding static variables shared between mapper and reducer classes.